stm32mp1: GPIO enhancement

etienne-lms opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented on core/arch/arm/plat-stm32mp1/main.c:
```diff
@@ -298,16 +298,25 @@ vaddr_t stm32mp_bkpreg(unsigned int idx)
 	assert(bank <= GPIO_BANK_K);
 
-	return io_pa_or_va(&gpios_nsec_base) + (bank * GPIO_BANK_OFFSET);
+	return io_pa_or_va(&x_base) + (bank * GPIO_BANK_OFFSET);
```
I wonder if `io_pa_or_va()` should rather be replaced with `io_pa_or_va_sec()` and `io_pa_or_va_nsec()`. I guess that for all invocations you know if you're targeting secure or non-secure IO-memory.

I'm not asking that we should change something here, it's just food for thought.

etienne-lms replied:
I agree that a generic `io_pa_or_va_{secure|nsec}()` would be nice.
@jforissier, should we keep the legacy `io_pa_or_va()`?

jforissier replied:
> @jforissier, should we keep the legacy `io_pa_or_va()`?

Not necessarily, if it's not needed we could as well remove it. Even if some use case comes later, we can always add it back.

etienne-lms replied:
I have created P-R https://github.com/OP-TEE/optee_os/pull/3020 for this.

Note that most optee_os drivers using `io_pa_or_va()` are UART console drivers. They can be used on secure or non-secure peripherals. Therefore I did not remove `io_pa_or_va()` as still used from these drivers.

jenswi-linaro commented on core/arch/arm/plat-stm32mp1/shared_resources.c:
```diff
@@ -139,11 +144,46 @@ static __maybe_unused const char *shres2str_state(enum stm32mp_shres id)
+
 static unsigned int get_gpioz_nbpin(void)
+{
+	if (gpioz_nbpin < 0) {
```
I would have preferred one function to read out and save the pin count and another to return it. That way it's clear when the driver is initializing, instead of being in a maybe half initialized state until `stm32mp1_init_shres()` has been run.

jenswi-linaro commented on core/arch/arm/plat-stm32mp1/shared_resources.c:
```diff
@@ -139,11 +144,46 @@ static __maybe_unused const char *shres2str_state(enum stm32mp_shres id)
+	if (gpioz_nbpin < 0)
+		panic();
+
+	return (unsigned int)gpioz_nbpin;
```
Is this cast really needed?

jenswi-linaro commented on core/arch/arm/plat-stm32mp1/main.c:
```diff
@@ -298,16 +298,25 @@ vaddr_t stm32mp_bkpreg(unsigned int idx)
+	/* Get non-secure mapping address for GPIOZ */
+	if (bank == GPIO_BANK_Z) {
+		if (!cpu_mmu_enabled())
+			return (vaddr_t)z_base.pa;
```
Is this cast needed?

jenswi-linaro commented on core/arch/arm/plat-stm32mp1/main.c:
```diff
@@ -298,16 +298,25 @@ vaddr_t stm32mp_bkpreg(unsigned int idx)
+			return (vaddr_t)z_base.pa;
+
+		if (!z_base.va)
+			z_base.va = (vaddr_t)phys_to_virt(z_base.pa,
```
Cast needed?

etienne-lms commented:
Rebased to fix merge conflict.
Commits related to secure/non-secure mapping of GPIO banks removed from this P-R since addressed in https://github.com/OP-TEE/optee_os/pull/3020.
Remains GPIOZ bank pins count (for clarity, I squashed the fixup commit for review comment) and PM.

jenswi-linaro commented on core/arch/arm/plat-stm32mp1/shared_resources.c:
```diff
@@ -173,10 +178,39 @@ static __maybe_unused const char *shres2str_state(enum stm32mp_shres id)
+	return gpioz_nbpin;
+}
+
+static TEE_Result set_gpioz_nbpin_from_dt(void)
```
Why isn't this called from `stm32mp1_init_shres()`?

etienne-lms replied:
`stm32mp1_init_shres()` is called at the end of the driver inits, since once called, driver can no more register shared peripheral (`registering_locked=true`), hence `driver_init_late(stm32mp1_init_shres);`.
Reading gpioz pin count from DT should be done before any driver manipulate GPIOs, hence before drivers init, hence `service_init(set_gpioz_nbpin_from_dt);`.

jenswi-linaro replied:
I see, thanks for the explanation. A comment explaining that here would be nice.

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`
